auction-server/src/opportunity/repository/models.rs

```diff
@@ -22,48 +37,67 @@ pub enum OpportunityRemovalReason {
+pub struct OpportunityMetadataEvm {
+    pub target_contract:   Address,
+    #[serde(with = "crate::serde::u256")]
+    pub target_call_value: U256,
+    pub target_calldata:   Bytes,
+}
+
+#[serde_as]
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct OpportunityMetadataSvmClientKamino {
+    #[serde_as(as = "Base64")]
+    pub order: Vec<u8>,
+}
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+#[serde(tag = "name", rename_all = "lowercase")]
+pub enum OpportunityMetadataSvmClient {
+    Kamino(OpportunityMetadataSvmClientKamino),
+}
+
+#[serde_as]
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct OpportunityMetadataSvm {
+    pub client:     OpportunityMetadataSvmClient,
+    #[serde_as(as = "DisplayFromStr")]
+    pub router:     Pubkey,
+    #[serde_as(as = "DisplayFromStr")]
+    pub permission: Pubkey,
+    #[serde_as(as = "DisplayFromStr")]
+    pub block_hash: Hash,
 }
 
-impl TryFrom<Opportunity> for entities::OpportunityEvm {
-    type Error = anyhow::Error;
+pub trait OpportunityMetadata:
+    std::fmt::Debug + Clone + Serialize + DeserializeOwned + Send + Sync + Unpin + 'static
+{
+    fn get_chain_type() -> ChainType;
+}
 
-    fn try_from(val: Opportunity) -> Result<Self, Self::Error> {
-        Ok(entities::OpportunityEvm {
-            core_fields:       entities::OpportunityCoreFields {
-                id:             val.id,
-                creation_time:  val.creation_time.assume_utc().unix_timestamp_nanos(),
-                permission_key: PermissionKey::from(val.permission_key),
-                chain_id:       val.chain_id,
-                sell_tokens:    serde_json::from_value(val.sell_tokens)
-                    .map_err(|e| anyhow::anyhow!(e))?,
-                buy_tokens:     serde_json::from_value(val.buy_tokens)
-                    .map_err(|e| anyhow::anyhow!(e))?,
-            },
-            target_contract:   ethers::abi::Address::from_slice(&val.target_contract),
-            target_call_value: val.target_call_value.to_string().parse().unwrap(),
-            target_calldata:   Bytes::from(val.target_calldata),
-        })
+impl OpportunityMetadata for OpportunityMetadataEvm {
+    fn get_chain_type() -> ChainType {
+        ChainType::Evm
     }
 }
 
-impl From<entities::OpportunityRemovalReason> for OpportunityRemovalReason {
-    fn from(reason: entities::OpportunityRemovalReason) -> Self {
-        match reason {
-            entities::OpportunityRemovalReason::Expired => OpportunityRemovalReason::Expired,
-            entities::OpportunityRemovalReason::Invalid(_) => OpportunityRemovalReason::Invalid,
-        }
+impl OpportunityMetadata for OpportunityMetadataSvm {
+    fn get_chain_type() -> ChainType {
+        ChainType::Svm
     }
 }
+
+#[derive(Clone, FromRow, Debug)]
+pub struct Opportunity<T: OpportunityMetadata> {
+    pub id:             Uuid,
+    pub creation_time:  PrimitiveDateTime,
+    pub permission_key: Vec<u8>,
```

Comment:
the permission seems to exist in `OpportunityMetadataSvm`, it should probably only exist in 1 place